### PR TITLE
avoid using invalid iterator when erasing items using an iterator in a loop

### PR DESCRIPTION
**Pull request by wjwwood**
Occurred when updating to `cppcheck` to 1.82 on macOS.

See: https://github.com/ros2/build_cop/issues/74#issuecomment-362724075

### CODE REVIEW

**Comment by wjwwood** [comment]
Here's a build on the updated macOS machine:

- macOS: [Build Status](http://ci.ros2.org/job/ci_osx/3265/)

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp [hunk 1]
```diff
@@ -78,7 +78,7 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
   {
     for (auto it = guard_conditions_.begin(); it != guard_conditions_.end(); ++it) {
       if (*it == guard_condition) {
-        guard_conditions_.erase(it);
+        it = guard_conditions_.erase(it);
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: Why is this necessary? The next line breaks the loop and leaves the scope of `it`. I would consider the newly assigned `it` unused.

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp [hunk 2]
```diff
@@ -266,11 +266,11 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
         }
         any_exec->callback_group = group;
         any_exec->node_base = get_node_by_group(group, weak_nodes);
-        subscription_handles_.erase(it);
+        it = subscription_handles_.erase(it);
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: Same here.

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp [hunk 3]
```diff
@@ -301,11 +301,11 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
         any_exec->service = service;
         any_exec->callback_group = group;
         any_exec->node_base = get_node_by_group(group, weak_nodes);
-        service_handles_.erase(it);
+        it = service_handles_.erase(it);
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: And here.

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp [hunk 4]
```diff
@@ -334,11 +334,11 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
         any_exec->client = client;
         any_exec->callback_group = group;
         any_exec->node_base = get_node_by_group(group, weak_nodes);
-        client_handles_.erase(it);
+        it = client_handles_.erase(it);
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: And here.

File: rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp (hunk 1 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp]: That's a good point, I over matched it, I think I fixed it up correctly here: e0be1f6

**Comment by wjwwood** [comment]
Windows with upgrade and latest changes:

- Windows: [Build Status](http://ci.ros2.org/job/ci_windows/4034/)
